src/lang-promql/lsp/client.ts

```diff
@@ -0,0 +1,61 @@
+import axios from 'axios';
+import { AutocompleteContext, Completion, CompletionResult } from "@codemirror/next/autocomplete";
+import { AutocompleteResponse, TextEdit } from "./model";
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function isIterable(obj: any): boolean {
+  return obj != null && typeof obj[ Symbol.iterator ] === 'function';
+}
+
+export class LSPClient {
+  private readonly url: string
+  private readonly autocompleteEndpoint = "/completion"
+
+  constructor(url: string) {
+    this.url = url;
+  }
+
+  autocomplete(context: AutocompleteContext): Promise<CompletionResult> {
+    const {state, pos} = context
+    const line = state.doc.lineAt(pos)
+    const body = {
+      expr: state.doc.sliceString(0),
+      limit: 100,
+      // positionChar start at 0 at the beginning of the line for LSP
+      positionChar: pos - line.from - 1 > 0 ? pos - line.from - 1 : pos - line.from,
+      // positionLine start at 0 from LSP and at 1 from CMN
+      positionLine: line.number - 1,
+    }
+    return axios.post<AutocompleteResponse[]>(this.url + this.autocompleteEndpoint, body)
+      .then((response) => {
+        const options: Completion[] = []
+        // for every textEdit present, they qll have exactly the same value
```

Comment:
These TextEdits aren't necessarily all the same.

For example, if a completion result is a function, the completion will include parenthesis and the TextEdit will instruct the client to place the cursor between those. For a metric name no such thing is done.

Reply:
that doesn't really matter actually. Here we are only considering the position `start` not the position `end`.
And the `start` position is always the same

Maybe I should change a bit my comment, since it's not super clear